### ISSUE TITLE
Type fields should store kinds instead of aliases

### ISSUE THREAD

**Issue by brendanzab** [gluon-lang/gluon#189]
Thoughts?

```diff
    ExtendRow {
        /// The associated types of this record type
-       types: Vec<Field<Id, Alias<Id, T>>>,
+       types: Vec<Field<Id, ArcKind>>,
        /// The fields of this record type
        fields: Vec<Field<Id, T>>,
        /// The rest of the row
        rest: T,
    },
```

**Comment by brendanzab** [comment]
Hmm, so it seems like the compiler uses the aliases to ship types around... :/ https://github.com/gluon-lang/gluon/blob/5904af005662aa7466b6ebb04a8b94a15469b92f/vm/src/compiler.rs#L784-L793

**Comment by Marwes** [comment]
I don't see how a kind would be enough. The alias needs to know which type is aliased to be able to unify with that type.

**Comment by brendanzab** [comment]
Shouldn't that be at the value level?

**Comment by Marwes** [comment]
Consider 
```f#
let y = { x = 0}
type Test a = { x : a }
let z : Test Int = y
1
```

For the `z` binding to unify we need to know which type `Test` aliases. We could also store this in a `Map` but then it becomes complicated when `Test` are defined in a different module

**Comment by brendanzab** [comment]
Shouldn't `FooA` and `FooB` be of the same type here?

```rust,fs
type Foo = { type x : Type }
let FooA = { type x = Int }
let FooB = { type x = String }
```

**Comment by brendanzab** [comment]
This is kind of a hairy part of the semantics where value/type/kind get a little confusing - might need to look over the semantics of other PLs here... 😟

**Comment by Marwes** [comment]
> Shouldn't FooA and FooB be of the same type here?

How can they be the same since they differ in types of the `x` field?

**Comment by brendanzab** [comment]
Hmm, true. I'm guessing that `Foo` should be trapped at the kind level, where they both would unify. Messed around in Lean to see what it does:

```lean
variable T : Type₁

structure foo :=
    mk :: (x : Type₁)

check ℕ                  -- ℕ : Type₁
check foo                -- foo : Type₂
check foo.mk ℕ           -- foo.mk ℕ : foo
check foo.mk T           -- foo.mk T : foo
check foo.mk (foo.mk T)  -- Error - found type `foo`, expected type `Type₁`
```

**Comment by Marwes** [comment]
Missed completely that the it was type fields! It should work the same though, two types with different type fields, either in the fields name or type (right-hand side) does not unify with each other. ML does it differently and allows the types to be different which essentially enables sub-typing.

>  I'm guessing that Foo should be trapped at the kind level

Yeah, that seems necessary as a minimum

**Comment by brendanzab** [comment]
> ML does it differently and allows the types to be different

ML doesn't allow types in records though. That's restricted to the module level. Or is that what you are talking about?

**Comment by Marwes** [comment]
Yeah was thinking about the module level, since we allow types in record like 1ML our records are closer to modules than ML records.

**Comment by brendanzab** [comment]
1ML has a concept of 'large' and 'small' types - dunno if that has anything to do with anything...

**Comment by Marwes** [comment]
1ML only allows subtyping for small types which is any type which does not employ `type` in itself (p.5, Predicativity in https://www.mpi-sws.org/~rossberg/1ml/1ml-extended.pdf). Not entirely sure how all the interactions are with it but intuitively `type` causes loss of information which means at that point it is not possible to check that the types are (structurally) the same.

**Comment by brendanzab** [comment]
Yeah, I was reading that part - took a while to digest though. This was useful:

> Notably, the case A with T1 literally implies `type type : ̸ type` (although `type type` itself is a well-formed expression!). The main challenge with first-class modules is preventing such a `type : type` situation, and the separation into a small universe (denoted by `type`) and a large one (for which no syntax exists) achieves that.
